docs/CONTRIBUTING.md

```diff
@@ -0,0 +1,513 @@
+
+## Refactoring
+
+* **Any PR that does a large-scale refactoring must be in its own PR**.  This includes PRs that touch multiple subsystems.  Refactoring often adds line noise that obscures the new functional changes that the PR proposes.  Small-scale refactorings are permitted to ship with functional changes.
+
+* Refactoring PRs can generally be bigger, because they are easier to review.  However, **large refactorings that could impact the functional behavior of the system should be discussed first** before carried out.  This is because it is imperative that they do not stay open for very long (to keep the submitter's maintenance burden low), but nevertheless reviewing them must still take at most 2 hours.  Discussing them first front-loads part of the review process.
+
+## Databases
+
+* If at all possible, **the database schema should be preserved**.  Exceptions can be made on a case-by-case basis.  The reason for this is that it's a big ask for people to re-sync nodes from genesis when they upgrade to a new point release.
+
+* Any changes to a database schema must also ship with a **new schema version and new schema migration logic**, as well as _test coverage_ for it.
+
+* The submitter must verify that **any new database columns are indexed**, as relevant to the queries performed on them.  Table scans are not permitted if they can be avoided (and they almost always can be).  You can find table scans manually by setting the environment variable `BLOCKSTACK_DB_TRACE` when running your tests (this will cause every query executed to be preceded by the output of `EXPLAIN QUERY PLAN` on it).
+
+* Database changes **cannot be consensus-critical** unless part of a hard fork (see below).
+
+* If the database schema changes and no migration can be feasibly done, then the submitter **must spin up a node from genesis to verify that it works** _before_ submitting the PR.  This genesis spin-up will be tested again before the next node release is made.
+
+## Data Input
+
+* **Data from the network, from Bitcoin, and from the config file is untrusted.**  Code that ingests such data _cannot assume anything_ about its structure, and _must_ handle any possible byte sequence that can be submitted to the Stacks node.
+
+* **Data previously written to disk by the node is trusted.** If data loaded from the database that was previously stored by the node is invalid or corrupt, it is appropriate to panic.
+
+* **All input processing is space-bound.**  Every piece of code that ingests data must impose a maximum size on its byte representation.  Any inputs that exceed this size _must be discarded with as little processing as possible_.
+
+* **All input deserialization is resource-bound.** Every piece of code that ingests data must impose a maximum amount of RAM and CPU required to decode it into a structured representation.  If the data does not decode with the allotted resources, then no further processing may be done and the data is discarded.
```

Comment:
Might be helpful to link to example(s) in the code of these patterns for reference

Reply:
Have a look at `src/net/http.rs` and how it uses `BoundReader` to prevent more than a fixed amount of data from being parsed before declaring the inbound message invalid.

Reply:
Added in 6eb9d0141